runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java

```diff
@@ -222,7 +282,117 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
     @Override
     public Map<PValue, ReplacementOutput> mapOutputs(
         Map<TupleTag<?>, PValue> outputs, WriteFilesResult<DestinationT> newOutput) {
-      return Collections.emptyMap();
+      return ReplacementOutputs.tagged(outputs, newOutput);
+    }
+  }
+
+  /**
+   * Flink has a known problem of unevenly assigning keys to key groups (and then workers) for cases
+   * that number of keys is not >> key groups. This is typical scenario when writing files, where
+   * one do not want to end up with too many small files. This {@link ShardingFunction} implements
+   * what was suggested on Flink's <a
+   * href="http://mail-archives.apache.org/mod_mbox/flink-dev/201901.mbox/%3CCAOUjMkygmFMDbOJNCmndrZ0bYug=iQJmVz6QvMW7C87n=pw8SQ@mail.gmail.com%3E">mailing
+   * list</a> and properly chooses shard keys in a way that they are distributed evenly among
+   * workers by Flink.
+   */
+  private static class FlinkAutoBalancedShardKeyShardingFunction<UserT, DestinationT>
+      implements ShardingFunction<UserT, DestinationT> {
+
+    private final int parallelism;
+    private final int maxParallelism;
+    private final Coder<DestinationT> destinationCoder;
+    private final Map<CacheKey, ShardedKey<Integer>> cache = new HashMap<>();
+    private final Set<Integer> usedSalts = new HashSet<>();
+
+    private int shardNumber = -1;
+
+    private FlinkAutoBalancedShardKeyShardingFunction(
+        int parallelism, int maxParallelism, Coder<DestinationT> destinationCoder) {
+      this.parallelism = parallelism;
+      this.maxParallelism = maxParallelism;
+      this.destinationCoder = destinationCoder;
+    }
+
+    @Override
+    public ShardedKey<Integer> assignShardKey(
+        DestinationT destination, UserT element, int shardCount) throws Exception {
+
+      if (shardNumber == -1) {
+        shardNumber = ThreadLocalRandom.current().nextInt(shardCount);
```

Comment:
Is there any benefit to starting with a random shard number as opposed to `0`?

Reply:
I just took it from WriteFiles, where it says
```
// We want to desynchronize the first record sharding key for each instance of
// ApplyShardingKey, so records in a small PCollection will be statistically balanced.
```